Update of parameter massager

espenfl opened the pull request:
I, the author consider this PR
 - [x] ready to be reviewed and merged (as soon as tests have passed)
 - [ ] work in progress (early feedback welcome)

## Description
When allowing selective dynamics flags in PR #408 some bugs were introduced which caused e.g. the flags to be not put correctly in the code namespace. This is fixed and in addition we introduce the following specification: the namespaces allowed on the input parameters are restricted. This means that if the user want sto override a parameter (send it straight to VASP) it needs to be put in the code namespace. If any parameter is detected on the input parameter root and it is not in the allowed namespace, an exception will be raised. While updating we also disconnected the massager from the workchain, which was previously supplied to get access to the exit codes. Now, we instead raise exceptions with a sensible text and then catch this in the VASP workchain and customize an exit code there that incorporates the exception message.

espenfl commented:
@atztogo @zhubonan Could you have a quick look at this before we merge. I think the changes should be for the better and also fixed a bug that was introduced in #408. I consider this the final PR before making a new release.

espenfl commented:
I also added this: https://github.com/aiida-vasp/aiida-vasp/wiki/Memo-on-passing-parameters

espenfl commented:
Also, since we change the way the `energies` are presented and how we parse parameters (now override have to be put in `inputs.parameters.code`) I suggest we increment major version.

atztogo reviewed (approved):
OK!

atztogo commented:
I agree to increment the major version number.

zhubonan commented:
Err  I realised that the `vasp` is renamed to `code` in the previous PR. I did not notice last month. 

This will introduce lots of compatibility issues for existing workchains. I am not sure about other users but I have written quite a few. Can we make `vasp` also a valid namespace? I think querying for `vasp.ibrion` makes more sense than `code.ibrion`.

 Also, the assistant is only capable of translating setting into VASP parameters anyway. I don't think there is any point to have `code` as the namespace to make it general. If one day one want to do this for QE they will just need to rename `vasp` to `qe` in their package. 

Otherwise, I think it is a good thing to have the logic of having all keys should reside in the namespace.

zhubonan commented on aiida_vasp/assistant/parameters.py:
```diff
@@ -12,22 +12,7 @@
-        'tetra': True
-    }
-}
+_ALLOWED_NAMESPACES = ['electronic', 'smearing', 'charge', 'dynamics', 'bands', 'relax', 'converge', 'code']
```
For backward compatibility, I would suggest we also add `vasp` as the allowed namespaces. 
Those from both `code` and `vasp` can be combined together for messaging use.

zhubonan commented on aiida_vasp/assistant/parameters.py:
```diff
@@ -187,38 +176,40 @@ def _load_valid_params(self):
-                    self._massage.code[key] = item
-                else:
-                    break
+                for key, item in self._parameters.code.items():
```
For backward compatibility:

Here, we can search for keys in both `code` and `vasp` namespace and combine them into self._message.code

espenfl replied:
We can for instance do this and issue a deprecation warning. Eventually I think it is for the better to come up with something else. It should be something that represents `incar`, but is not called that of course. See other reply.

zhubonan commented on aiida_vasp/workchains/relax.py:
```diff
@@ -313,9 +314,12 @@ def verify_next_workchain(self):
+            # Make sure at the very minimum we attach the misc node (if it exists) that contains notifications and other
             # quantities that can be salvaged
-            self.out('misc', workchain.outputs['misc'])
+            try:
```
Not sure if try..except is necessary - `if 'misc' in workchain.outputs` should be fine. 
If no misc output something is very wrong....

espenfl replied:
The reason this is here is that if something is really wrong we would like to have  a clean report where you see the error and not a bunch of errors related to missing `misc` (which is not the true error).

zhubonan reviewed (commented):
Thanks for the effort! This should make the calculations less prone to errors.
My main comment is to include `vasp` as the valid input for backward compatibility.

Also, I think `vasp` is a more meaningful name for incar tags. We can keep both `vasp` and `code` and possible ways to override actual parameters.

espenfl commented:
> Err I realised that the `vasp` is renamed to `code` in the previous PR. I did not notice last month.
> 
> This will introduce lots of compatibility issues for existing workchains. I am not sure about other users but I have written quite a few. Can we make `vasp` also a valid namespace? I think querying for `vasp.ibrion` makes more sense than `code.ibrion`.
> 
> Also, the assistant is only capable of translating setting into VASP parameters anyway. I don't think there is any point to have `code` as the namespace to make it general. If one day one want to do this for QE they will just need to rename `vasp` to `qe` in their package.
> 
> Otherwise, I think it is a good thing to have the logic of having all keys should reside in the namespace.

Okey, so the thought behind this is the following. All the higher lying workchains should not contain code specific logic, variable names etc. Since we call `inherit_and_merge_parameters` in these workchains, they needed to be stripped of the `vasp` keyword. If you use several codes in a workflow you would typically do `mycode.parameters.code` or `parameters.mycode.code` etc. Thus, it would also be a bit weird to do `vasp.paratemers.vasp` or `parameters.vasp.vasp`.

We thought that `code` was maybe a better substitute, but I agree, it is not optimal, especially since we already use the `Code` object typically as `code`. I am happy for other suggestions, but I think we should try to keep it non code specific.

Also remember that this was also introduced since we now have a `dynamics` (could be others as well) flag, so what is going into the calculation is `parameters.code` and `parameters.dynamics` etc. Here it would also be a bit strange to use `parameters.vasp` instead of `parameters.code` (but again, `code` is certainly not optimal).

`incar` would obviously be what comes to mind, but we should definitively stay away from that for the same reason as `vasp`, but also in addition to the fact that it might be removed so that a single file is used in the future.

A few suggestions (if we want to go with something else than `code`): `inputs`, `raw`, `direct`, `input_parameters` @zhubonan @atztogo @JPchico What do you think?

atztogo commented:
Naming variables is one of the most important parts in coding. For this, I know there are many discussion, but if we can not determine names right now, I suggest some names that are 'grep'able. At least `code` is not good in terms of 'grep'ability. For some simple names like `code`, `node`, etc in AiiDA terminology, those should be considered as 'reserved' words that shoud not be used for the other purpose (type). Do we have such a list of semi-reserved words in AiiDA?

espenfl commented:
> Naming variables is one of the most important parts in coding. For this, I know there are many discussion, but if we can not determine names right now, I suggest some names that are 'grep'able. At least `code` is not good in terms of 'grep'ability. For some simple names like `code`, `node`, etc in AiiDA terminology, those should be considered as 'reserved' words that shoud not be used for the other purpose (type). Do we have such a list of semi-reserved words in AiiDA?

Supported. But, when we have and use namespaces that should also be accounted for while grep'ing, leaving us more flexibility. Since our parameters are sort of verified, or processed, maybe we can utilize that in the naming scheme? So we could do something like `verified_inputs` or similar. We still have the `dynamics` key however.

JPchico commented:
> > Naming variables is one of the most important parts in coding. For this, I know there are many discussion, but if we can not determine names right now, I suggest some names that are 'grep'able. At least `code` is not good in terms of 'grep'ability. For some simple names like `code`, `node`, etc in AiiDA terminology, those should be considered as 'reserved' words that shoud not be used for the other purpose (type). Do we have such a list of semi-reserved words in AiiDA?
> 
> Supported. But, when we have and use namespaces that should also be accounted for while grep'ing, leaving us more flexibility. Since our parameters are sort of verified, or processed, maybe we can utilize that in the naming scheme? So we could do something like `verified_inputs` or similar. We still have the `dynamics` key however.
> 
> I do not know of such a list. Also, these are typically instances of say the `Code` etc.

I agree with what you have raised, indeed `code` is not the best solution, I think that `vasp` is also not the best due to the issues that you have named. 

I think that one should be a bit explicit in how to call this, specially as there are several namespaces and it can get confusing quite quickly. As much as possible the name should specify that these are inputs related to the `incar`. I understand why one does not want to use the `incar` tag, but perhaps something that alludes to the fact that these are the kind of parameters that one is referring to. Perhaps something like `code_specific`, `calculation_control` or something similar.

espenfl commented:
> > > Naming variables is one of the most important parts in coding. For this, I know there are many discussion, but if we can not determine names right now, I suggest some names that are 'grep'able. At least `code` is not good in terms of 'grep'ability. For some simple names like `code`, `node`, etc in AiiDA terminology, those should be considered as 'reserved' words that shoud not be used for the other purpose (type). Do we have such a list of semi-reserved words in AiiDA?
> > 
> > 
> > Supported. But, when we have and use namespaces that should also be accounted for while grep'ing, leaving us more flexibility. Since our parameters are sort of verified, or processed, maybe we can utilize that in the naming scheme? So we could do something like `verified_inputs` or similar. We still have the `dynamics` key however.
> > I do not know of such a list. Also, these are typically instances of say the `Code` etc.
> 
> I agree with what you have raised, indeed `code` is not the best solution, I think that `vasp` is also not the best due to the issues that you have named.
> 
> I think that one should be a bit explicit in how to call this, specially as there are several namespaces and it can get confusing quite quickly. As much as possible the name should specify that these are inputs related to the `incar`. I understand why one does not want to use the `incar` tag, but perhaps something that alludes to the fact that these are the kind of parameters that one is referring to. Perhaps something like `code_specific`, `calculation_control` or something similar.

Yes, we should also look at this in light of the other namespaces, e.g. `dynamic` (which we might also change before making the release). I think it would be wise that we reach some acceptable terminology here in order to not have to do this change again (unless there is strict need for it). Ideally, maybe the `dynamic` should also be something that we put into the existing `code` as a namespace and then this is untangled in the calculation. Let us think about this a bit. I will try to make a suggestion today.

zhubonan commented:
> Okey, so the thought behind this is the following. All the higher lying workchains should not contain code specific logic, variable names etc. Since we call `inherit_and_merge_parameters` in these workchains, they needed to be stripped of the `vasp` keyword. If you use several codes in a workflow you would typically do `mycode.parameters.code` or `parameters.mycode.code` etc. Thus, it would also be a bit weird to do `vasp.paratemers.vasp` or `parameters.vasp.vasp`.

I think this is probably down to the design of the workflows - I personally don't use the `inherit_and_merge_parameters` way but utilises the exposed inputs when designing the workchains. I am not so sure if we should emphasis too much for generalisability for now. 

My vote for `vasp` is that it is descriptive to what it does at the `VaspWorkChain` level - it is the namespace where VASP specific tags go. If someone ever wants to use the stack to run QE calculations there are changes to the parameters massager needed anyway to get it to work, then QE specific tags can be put into `qe`.

espenfl commented:
Okey, here is my suggestion. I believe we should rely on the functionality of the massager. Main role of that is to translate to VASP parameter space.

After the massager, the namespaces in `parameters` should be ready for the calculation. It is tempting to use the file names for now (see *). This means that all parameters that should go into INCAR sits in `parameters.incar`, POSCAR, in `parameters.poscar` etc.  (see **). The massager is executed in the `VaspWorkChain` which is the recommended entry point. 

If someone wants to call the calculation directly they would just have to adhere to these namespaces. Similarly, this pass through should work from any workchain (but should never be mentioned in the higher workchains itself as these should strive to be code agnostic). In order to make this easier and more explicit I suggest we put overrides in say `parameters.override` or `parameters.passthrough` such that e.g. INCAR related flags are put in `parameters.override.incar.mytag` etc. Doing it like this makes it possible to always pass through the `override` namespace and also check for unsupported tags on `parameters` root so that users do not set parameters that they believe should be accepted and set in INCAR. After the massager the `override` is removed and the final parameters sits in `parameters.incar` etc.

For the namespaces before the massager, those are already defined, e.g. `smearing`, `electronic` etc. We only need to pass the `override` namespace.

* It is tempting to avoid `incar` etc. as namespaces, but also convenient with respect to existing VASP users. If VASP changes in the future (say just a single HDF file) we would have to change this to make it compliant. But we could also imaging supporting both strategies for a while, which would most likely also be the case it will be done. Then we would e.g. have `parameters.hdf5` and `parameters.override.hdf5`.
** With this a new question appear. How closely connected should the content of the container and the content in the file be. For `parameters.incar` it would be 1:1, for `parameters.poscar` only partly (POSCAR will be a combination of `parameters.poscar.positions_dof` and the supplied `inputs.structure` to the calculation). And for KPOINTS, it will be fully spec'ed from the `inputs.kpoints`. But it is not possible to make this more compact as we have already discussed that the `postitions_dof` should not be part of `StructureData` or something like `GeneralStructureData` etc., meaning we need anyway a second container, like `parameters.poscar.positions_dof`. And I think this is fine as long as we document it and are clear about how it works.

I am happy for suggestions for the `parameters.incar`, `parameters.poscar` etc. but I fail to come up with something useful, without making it bike shedding. So I would maybe suggest to use this as it would maybe also be a benefit for new `aiida-vasp`, but experienced VASP users that only want to run `aiida-vasp` without too much workflow at first.

espenfl commented:
> > Okey, so the thought behind this is the following. All the higher lying workchains should not contain code specific logic, variable names etc. Since we call `inherit_and_merge_parameters` in these workchains, they needed to be stripped of the `vasp` keyword. If you use several codes in a workflow you would typically do `mycode.parameters.code` or `parameters.mycode.code` etc. Thus, it would also be a bit weird to do `vasp.paratemers.vasp` or `parameters.vasp.vasp`.
> 
> I think this is probably down to the design of the workflows - I personally don't use the `inherit_and_merge_parameters` way but utilises the exposed inputs when designing the workchains. I am not so sure if we should emphasis too much for generalisability for now.
> 
> My vote for `vasp` is that it is descriptive to what it does at the `VaspWorkChain` level - it is the namespace where VASP specific tags go. If someone ever wants to use the stack to run QE calculations there are changes to the parameters massager needed anyway to get it to work, then QE specific tags can be put into `qe`.

I think `vasp` is descriptive, but not as a substitute for `incar`. When using `vasp` I would also expect the `positions_dof` to be in there somehow. Also, we would probably use `vasp` as a code container when using multiple codes, e.g. VASP and Wannier90, which is pretty common and something users do rather often.

espenfl commented:
> My vote for `vasp` is that it is descriptive to what it does at the `VaspWorkChain` level - it is the namespace where VASP specific tags go. If someone ever wants to use the stack to run QE calculations there are changes to the parameters massager needed anyway to get it to work, then QE specific tags can be put into `qe`.

One should in principle only have to use a `QEWorkChain` which will call the relevant massager, which would be up to the QE folks to maintain.

zhubonan commented:
Anyway, I think we just have to decide a name and possibly don't touch it ever again.

My concern about `overrides` is that one will typically use default VASP settings and some tags are always present (like `PREC`, `NCORE` etc), those are just routines rather than one-off overrides. At the `CalcJob` level everything in `parameters` should actually all goes into the INCAR files to keep it simple. 

I realised that the previous selective dynamics PR also changed this in `calcs/vasp.py` so that a `code` namespace is added at `VaspCalculation` level. I think this is really unnecessary - this is an additional feature and should be dispatched using an additional non-required input port/or additional keys of the existing port (such as `settings`) rather than changing the whole schema of `parameters` at the level. Now `calcjob.get_builder_restart()` will give a build that does not work! 

My suggestions are that we keep `vasp/overrides/incar` at valid top-level keys for settings the incar tags. Since the massager does not do anything to the selective dynamics tags at all, why don't we just put it in settings like `settings['positions_dof']`, which then gets passed through to the `CalcJob`?

espenfl commented:
@atztogo and @JPchico: I discussed this a bit with @zhubonan and here is what we agreed on:

- We make a new input node on the `VaspCalculation` which houses the `positions_dof`. The massager would handle the setting of that input, which would be something like `inputs.positions_dof`.
- After the massager we remove the `code` namespace and again make sure the `parameters` are clean and only contain the tags that should go into `INCAR`.
- We rename the `code` keyword before the massager to `incar`. This means that if the user want to set any override flags, they would set this in `parameters.incar`. The massager would combine all inputs and make sure the valid tags that goes into INCAR is put on `inputs.parameters` after the massager completes.

With this approach, we make it more clear to the beginner where to supply custom tags and we keep the existing backwards compatibility of the `VaspCalculation`.

There is one _problem_ with this approach. We would potentially have to handle the `incar` key in the higher lying workchains (where VASP specific stuff should not reside). It would then be tempting to let all flags pass on `parameters` into the massager, but I would like to only allow certain namespaces to be valid (to avoid users setting `parameters.vasp.ibrion`, `parameters.ibrion` etc. and expect it to work, which is dangerous when VASP offers default settings). So what I will do is that I will make it possible to pass optional namespaces that will be passed directly to the massager in `settings`. `incar` would be supplied when using `aiida-vasp`. If a user wants another plugin, they can add flags, or let that plugin add it and we would only have to pass our defined namespaces (`_ALLOWED_NAMESPACES`) plus `_CODE_OVERRIDE_NAMESPACES` or something like that. Thus we do not need to touch the name `incar` anywhere.

Does this sound okey?

atztogo commented:
I can't follow your discussion without using the experimental implementation. Maybe we don't need to rush releasing. Or we can release, too, then work more. It's just the matter of incrementing the major version number. What is the priority?

espenfl commented:
> I can't follow your discussion without using the experimental implementation. Maybe we don't need to rush releasing. Or we can release, too, then work more. It's just the matter of incrementing the major version number. What is the priority?

I think we should not delay this particular release more than necessary as it contains a few important bug fixes. If you find it easier, I can update the PR to reflect the text.

atztogo commented:
Then, it's clear. We should just release the current code right now. Please postpone the further updates.

espenfl commented:
What is develop now since the last release would break backwards compatibility, so I would like to suggest that we do not do that. This was also the concern of @zhubonan. What is suggested above would not. At least not for the `VaspCalculation` interface.

zhubonan commented:
Hi @atztogo, here is a summary of the changes me and @espenfl discussed. 
 

### Last release 

- Before the massager (e.g. input to the `VaspWorkChain`) : `paramaters.vasp` holds the INCAR tags, although it is also possible to supply the INCAR tags at the top level.  
```
{
'vasp': {'prec': accurate', 'istart': 0}, ...
}
```
- After the massager (e.g. passed to `VaspCalculation`): `parameters` holds the INCAR tags at the top level, packed into a `Dict` and passed to the `VaspCalculation`'s `parameters` input port. 

```
{
'prec': accurate', 'istart': 0,
}
```

### Current development branch

#408 introduced changes to allow selective dynamics tags to be passed:

- Before the massager:  `parameters.code` holds the INCAR tags, and `parameters.dynamics` holds the selective dynamics input. 
```
{
'code': {prec': accurate', 'istart': 0}, 'dynamics': {....}, ...
}
```

- After the massager: `parameters.code` holds the INCAR tags and `parameters.dynamics.dynamics` holds the selective dynamics input used for writing the POSCAR. The rationale for doing this that additional input can be passed without added another input node at the `CalcJob` level. However, these changes are not backwards compatible and introduce changes for the data provenance. For example, calling `calcjob.get_builder_restart` will not give a working `builder` for restarting a calculation before the change was applied.

```
{
'code': {prec': accurate', 'istart': 0}, 'dynamics': {...},
}
```

### Proposed change 

We revert the change of the input format in #408 by keeping `parameters` input for `VaspCalculation` exactly the same as the previously released version. The selective dynamics will be introduced as a non-required input port without a default value, so not supplying it is just fine. Now, the massager behaviour is:

- Before the massager: `parameters.incar` holds the INCAR tags, and `parameters.dynamics` holds the selective dynamics input. Previously this namespace is called `parameters.vasp`. Renaming it to `incar` makes it clearer what it does (manually supplying INCAR tags).  @espenfl We can make `vasp` still valid, but issue deprecation warnings for now, then remove it later.

```
{
'incar': {prec': accurate', 'istart': 0}, 'dynamics': {...}, ....
}
```

- After the massager: `parameters` holds the INCAR tags at the top level just behave the last released version. The `dynamics` part in the `parameters` passed to `VaspWorkChain`, is separated and supplied as an additional input node, if required. 

```
{
prec': accurate', 'istart': 0,
}
```

+ another node dedicated for dynamics settings: 
```
{
'positions_dof': [....], 
}
```

-----

The other change is that previously, the massager also allows `parameters` contain top-level INCAR tags that get passed through, this is will not be allowed anymore after this PR. 

The reason for not just releasing another version using the current development branch is that it already contains #408 and introduces backwards-incompatible changes. Although we can say we break the compatibility by increasing the major version, this is unnecessary and can be avoided if the proposed change is merged before making the release. Otherwise, we need to increase the version number again to apply the proposed changes.

espenfl commented:
Yes, I will submit the PR tomorrow to illustrate the specific changes/reverts.

atztogo commented:
Thanks @zhubonan  for your detailed description. It became more clear for me now. For me, breaking backward compatibility is a minor issue, but I believe @zhubonan's suggestion is very good if I assume addition work needed is not that much.

> I think we should not delay this particular release more than necessary as it contains a few important bug fixes. 

@espenfl 's first priority is the release. This is what we have to respect the best. We should minimize the time to release. It may be nice to make the schedule to release to define until when and what versions will be released, for which we can use github projects (Kanban system) and wiki. Maybe hopefully, the last minor version would be released before the major version with @zhubonan 's request including DeprecationWarning if having backward compatibility as @zhubonan wrote.

Anyway, for me any implementation is OK. My comment is minor and ignorable. Including me in this discussion is the source of delay to release. So please go ahead as you want.

JPchico commented:
@espenfl and @zhubonan  I think that what you suggest is a really good idea. I wonder about bloat in the `parameters` entry after the massager has acted, in principle the user should not have to deal with this themselves, but it is something that I think one should keep in mind. Otherwise I think that is seems great.

I wonder about the bug that was introduced in #408 , it went undetected by the tests and in my own runs, so I think it might be a good idea to set a test to handle that kind of problem.

espenfl commented:
> I wonder about the bug that was introduced in #408 , it went undetected by the tests and in my own runs, so I think it might be a good idea to set a test to handle that kind of problem.

Yes, I indeed added a test so that hopefully if we change something like this in the future it would not go undetected. It only showed up when running e.g. the `run_relax` example, it did not relax since the `relax` namespace was not passed along.

espenfl commented:
@atztogo @zhubonan @JPchico I now updated the PR reflecting the changes. I did a force push with the updated changes in order to keep the commit history reasonably clean. All tests pass and workchains seem to work as expected. Please have a look if you want. I also simplified and added some additional docstrings, particularly in the parameter massager to make things, maybe more clear. I should also update the wiki to reflect the updates, but will do so when it is merged and we have settled the way of operation.

atztogo commented:
Thanks @espenfl . Merging PR #419 is my concern but it went OK.

zhubonan reviewed (approved):
Thank for the update! Looks superb to me.

espenfl commented:
Okey, thanks a lot for the contributions everyone. I will now merge it and continue with the release cycle.
